core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java

```diff
@@ -92,49 +99,121 @@ public ListenableFuture<Void> execute(
+                    session,
+                    plannerContext,
+                    accessControl,
+                    bindParameters(statement, parameters),
+                    true);
+
+            ColumnMetadata column = ColumnMetadata.builder()
+                    .setName(columnName.getValue())
+                    .setType(type)
+                    .setNullable(element.isNullable())
+                    .setComment(element.getComment())
+                    .setProperties(columnProperties)
+                    .build();
+
+            plannerContext.getMetadata().addColumn(session, tableHandle, qualifiedTableName.asCatalogSchemaTableName(), column);
         }
-        if (!element.isNullable() && !plannerContext.getMetadata().getConnectorCapabilities(session, catalogHandle).contains(NOT_NULL_COLUMN_CONSTRAINT)) {
-            throw semanticException(NOT_SUPPORTED, element, "Catalog '%s' does not support NOT NULL for column '%s'", catalogHandle, element.getName());
+        else {
+            accessControl.checkCanAlterColumn(session.toSecurityContext(), qualifiedTableName);
+
+            if (!element.isNullable()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with NOT NULL constraint is unsupported");
+            }
+            if (!element.getProperties().isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with column properties is unsupported");
+            }
+            if (element.getComment().isPresent()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with COMMENT is unsupported");
+            }
+            if (!columnHandles.containsKey(columnName.getValue().toLowerCase(ENGLISH))) {
+                throw semanticException(COLUMN_NOT_FOUND, statement, "Column '%s' does not exist", columnName);
+            }
+
+            List<String> parentPath = statement.getColumn().getField().getOriginalParts().subList(0, statement.getColumn().getField().getOriginalParts().size() - 1).stream()
+                    .map(Identifier::getValue)
+                    .collect(toImmutableList());
+            List<String> fieldPath = statement.getColumn().getField().getOriginalParts().subList(1, statement.getColumn().getField().getOriginalParts().size()).stream()
+                    .map(Identifier::getValue)
+                    .collect(toImmutableList());
+
+            ColumnMetadata columnMetadata = plannerContext.getMetadata().getColumnMetadata(session, tableHandle, columnHandles.get(columnName.getValue().toLowerCase(ENGLISH)));
+            Type currentType = columnMetadata.getType();
+            int fieldDepth = 0;
+            for (String fieldName : fieldPath) {
+                fieldDepth++;
+                if (currentType instanceof RowType rowType) {
+                    List<RowType.Field> candidates = rowType.getFields().stream()
+                            // case-sensitive match
```

Comment:
Row field references are resolved case-insensitive in expressions. We should do the same here for consistency.

Reply:
How can we handle the table having field paths with different cases? e.g. adding c field to under a.B in `a row(b row(...), B row(...))`

Reply:
If you have a column `a` with type `row(b row(...), B row(...))`, you can't select `a.B`, or any deeper nested field. This is because the `ExpressionAnalyzer` resolves row fields case-insensitive, and finds fields `b` and `B` ambiguous.

My point here is that we should remain consistent in field resolution. That would mean that you cannot add a field under `a.B`. To make sure we use consistent mechanism for field resolution, we could try resolve the `fieldPath` in `StatementAnalyzer`.

Another possible approach would be using the proper SQL identifier semantics. However, there are two issues with this approach:
- the column name is resolved case-insensitive, and only then we switch to another identifier semantics to resolve the remaining path. It might be confusing to the user. The current approach also has this drawback.
- we can canonicalize identifiers on engine side, but then the connectors should be ready to handle them accordingly to the semantics used by the underlying database. This is a complex change, tracked here: https://github.com/trinodb/trino/issues/17

Reply:
> If you have a column a with type row(b row(...), B row(...)), you can't select a.B, or any deeper nested field.

I was missing this fact. Then, case-insensitive makes sense. Thanks for explaining that. I will update this PR.